mesonbuild/modules/rust.py

```diff
@@ -189,6 +193,71 @@ def test(self, state: ModuleState, args: T.Tuple[str, BuildTarget], kwargs: Func
 
         return ModuleReturnValue(None, [new_target, test])
 
+    @typed_pos_args('rust.doctest', str, BuildTarget)
+    @typed_kwargs(
+        'rust.doctest',
+        *TEST_KWS,
+        DEPENDENCIES_KW,
+        LINK_WITH_KW,
+        KwargInfo(
+            'rust_args',
+            ContainerTypeInfo(list, str),
+            listify=True,
+            default=[],
+        ),
+        KwargInfo('is_parallel', bool, default=False),
+    )
+    def doctest(self, state: ModuleState, args: T.Tuple[str, T.Union[SharedLibrary, StaticLibrary]], kwargs: FuncTest) -> ModuleReturnValue:
+        name, base_target = args
+
+        # Link the base target's crate into the tests
+        kwargs['link_with'].append(base_target)
+        kwargs['depends'].append(base_target)
+        workdir = kwargs['workdir']
+        kwargs['workdir'] = None
+        new_target, tkwargs = self.test_common(state, args, kwargs)
+
+        # added automatically by rustdoc; keep things simple
+        tkwargs['args'].remove('--test')
+
+        # --test-args= is "parsed" simply via the Rust function split_whitespace().
+        # This means no quoting nightmares (pfew) but it also means no spaces.
+        # Unfortunately it's pretty hard at this point to accept e.g. CustomTarget,
+        # because their paths may not be known.  This is not a big deal because the
+        # user does not control the test harness, so make things easy and allow
+        # strings only.
+        if tkwargs['args']:
+            for i in tkwargs['args']:
+                if not isinstance(i, str):
+                    raise InvalidArguments('only strings are allowed in args for rust.doctest')
+                if len(i.split()) > 1:
+                    raise InvalidArguments('spaces are not allowed in args for rust.doctest')
+            tkwargs['args'] = ['--test-args=' + ' '.join(T.cast(T.Sequence[str], tkwargs['args']))]
```

Comment:
Why do we need the cast here?

Reply:
Because otherwise mypy thinks `tkwargs['args']` can also contain ExternalPrograms, CustomTargets, or BuildTargets, and you cannot do a `join`. In turn building a new list inside the `for` is complicated by the fact that lists are invariant - so you cannot assign a `list[str]` to a `list[str | ExternalProgram | CustomTarget | BuildTarget]`. This was the best I could come up with.

Reply:
casts should really be quoted though:
```suggestion
            tkwargs['args'] = ['--test-args=' + ' '.join(T.cast('T.Sequence[str]', tkwargs['args']))]
```

Aside: Avoid using typing.Sequence with strings at all costs. This says that it is legal for

```python
tkwargs['args'] = 'raw string'
```

since a str object is, itself, a sequence type where sequence operations treat it as a list of str objects in the form: `['r', 'a', 'w', ' ', 's', 't', 'r', 'i', 'n', 'g']`

Thus, you get:

```python
>>> ' '.join('raw string')
'r a w   s t r i n g'
>>> ' '.join(['raw string'])
'raw string'
```

Reply:
Done. However, it's not possible to use anything but Sequence here, because only sequences are covariant. :(